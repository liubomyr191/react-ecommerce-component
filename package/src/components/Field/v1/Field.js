import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import isEmpty from "lodash.isempty";
import { addTypographyStyles, applyTheme } from "../../../utils";

function applyValidationColor(themeProp) {
  return (props) => {
    let status;
    if (props.errors && props.errors.length) {
      status = "error";
    } else if (props.hasBeenValidated && props.value && props.value.length) {
      status = "success";
    } else if (props.inputFocused || props.buttonFocused) {
      status = "focus";
    } else {
      status = "default";
    }
    return applyTheme(`${themeProp}_${status}`);
  };
}

const StyledField = styled.div`
  margin-bottom: ${applyTheme("Field.spacingBelow")};
  margin-left: 0;
  margin-right: 0;
  margin-top: ${applyTheme("Field.spacingAbove")};
`;

const StyledLabel = styled.label`
  ${addTypographyStyles("FieldLabel", "labelText")}
  color: ${applyValidationColor("FieldLabel.color")};
  display: block;
  margin-bottom: ${applyTheme("FieldLabel.spacingToInput")};
  margin-left: 0;
  margin-right: 0;
  margin-top: 0;
`;

const StyledHelpText = styled.span`
  ${addTypographyStyles("FieldHelp", "labelText")}
  display: block;
  margin-bottom: 0;
  margin-left: 0;
  margin-right: 0;
  margin-top: ${applyTheme("FieldHelp.spacingToInput")};
`;

class Field extends Component {
  static isFormField = true;

  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    errors: PropTypes.array,
    helpText: PropTypes.string,
    isOptional: PropTypes.bool,
    isRequired: PropTypes.bool,
    label: PropTypes.node,
    labelClassName: PropTypes.string,
    labelFor: PropTypes.string
  };

  static defaultProps = {
    className: undefined,
    errors: undefined,
    label: undefined,
    labelClassName: undefined,
    labelFor: undefined,
    isOptional: false,
    isRequired: false
  };

  getClassName() {
    const { className, errors, isRequired } = this.props;
    const errorClass = Array.isArray(errors) && errors.length > 0 ? "has-error" : "";
    const requiredClass = isRequired ? "required" : "";
    return `${className || ""} ${errorClass} ${requiredClass}`.trim();
  }

  renderLabel() {
    const { errors, label, labelClassName, labelFor, isOptional } = this.props;

    return (
      <StyledLabel className={labelClassName} errors={errors} htmlFor={labelFor}>
        {label}
        {isOptional ? " (Optional)" : null}
      </StyledLabel>
    );
  }

  renderHelpText() {
    const { helpText } = this.props;
    return <StyledHelpText>{helpText}</StyledHelpText>;
  }

  render() {
    const { children, helpText, label } = this.props;
    return (
      <StyledField className={this.getClassName()}>
        {!isEmpty(label) && this.renderLabel()}
        {children}
        {!isEmpty(helpText) && this.renderHelpText()}
      </StyledField>
    );
  }
}

export default Field;
