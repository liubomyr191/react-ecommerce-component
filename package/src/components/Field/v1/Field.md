### Overview

Fields can be required, with more than one line, and have help text and labels.

```jsx
<div>
  <Field name="example" label="Label" helpText="Help text">
    <TextInput name="example" placeholder="This field has help text" />
  </Field>
  <Field name="example" label="Label" helpText="Help text">
    <TextInput name="example" placeholder="This field has help text" shouldAllowLineBreaks />
  </Field>
</div>
```

#### Usage

##### Optional Field
```jsx
<Field name="example" label="Label" isOptional>
  <TextInput name="example" placeholder="This field has help text" />
</Field>
```
