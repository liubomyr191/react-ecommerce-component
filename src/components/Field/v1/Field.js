import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import isEmpty from 'lodash.isempty';
import { applyTheme } from "./../../../helpers";

const StyledDiv = styled.div`
  color: #333333;
`;

class Field extends Component {
  static isFormField = true;

  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    errors: PropTypes.array,
    isRequired: PropTypes.bool,
    label: PropTypes.node,
    labelClassName: PropTypes.string,
    labelFor: PropTypes.string
  };

  static defaultProps = {
    className: undefined,
    errors: undefined,
    label: undefined,
    labelClassName: undefined,
    labelFor: undefined,
    isRequired: false
  };

  getClassName() {
    const { className, errors, isRequired } = this.props;
    const errorClass = Array.isArray(errors) && errors.length > 0 ? 'has-error' : '';
    const requiredClass = isRequired ? 'required' : '';
    return `${className || ''} ${errorClass} ${requiredClass}`.trim();
  }

  renderLabel() {
    const { label, labelClassName, labelFor } = this.props;

    return (
      <label className={labelClassName} htmlFor={labelFor}>{label}</label>
    );
  }

  render() {
    const { children, label } = this.props;

    return (
      <div className={this.getClassName()}>
        {!isEmpty(label) && this.renderLabel()}
        {children}
      </div>
    );
}
}

export default Field;
