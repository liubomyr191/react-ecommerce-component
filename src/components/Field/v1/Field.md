### Field Overview

```jsx
<div>
  <Field name="example" label="Text Field">
    <TextInput name="example" placeholder="Hint" />
  </Field>
  <Field name="example" label="Text Field">
    <TextInput name="example" placeholder="Hint" />
  </Field>
</div>
```
