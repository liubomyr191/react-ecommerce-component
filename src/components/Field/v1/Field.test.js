import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Field from "./Field";

test("basic snapshot", () => {
  const component = renderer.create(<Field />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
