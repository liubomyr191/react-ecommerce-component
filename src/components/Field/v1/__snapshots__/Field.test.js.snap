// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with label 1`] = `
.c0 {
  background-color: #fafafa;
  border: 1px solid #cccccc;
  border-radius: 0.125rem;
  color: #737373;
  crsor: pointer;
  font-family: inherit;
  font-size: 0.875rem;
  line-height: 1;
  padding: 0.75rem 0.625rem;
  width: 100%;
}

.c0::-webkit-input-placeholder {
  color: #cccccc;
}

.c0::-moz-placeholder {
  color: #cccccc;
}

.c0:-ms-input-placeholder {
  color: #cccccc;
}

.c0::placeholder {
  color: #cccccc;
}

.c0:focus {
  border-color: #8ce0c9;
}

.c0:-moz-read-only {
  color: #bfbfbf;
}

.c0:read-only {
  color: #bfbfbf;
}

.c1 {
  color: #737373;
  fill: currentColor;
  font-size: 0.875rem;
  height: 1rem;
  position: absolute;
  right: 0.75rem;
  top: 0.75rem;
  width: 1rem;
}

.c1 * {
  display: inline-block;
}

<div
  className=""
>
  <label
    className={undefined}
    htmlFor={undefined}
  >
    Foo
  </label>
  <p>
    Text above
  </p>
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <input
      className="c0"
      maxLength={undefined}
      name="test"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      placeholder={undefined}
      readOnly={false}
      type="text"
      value=""
    />
    <div
      className="c1"
      onClick={[Function]}
      value=""
    />
  </div>
  <p>
    Text below
  </p>
</div>
`;

exports[`renders with no label 1`] = `
.c0 {
  background-color: #fafafa;
  border: 1px solid #cccccc;
  border-radius: 0.125rem;
  color: #737373;
  crsor: pointer;
  font-family: inherit;
  font-size: 0.875rem;
  line-height: 1;
  padding: 0.75rem 0.625rem;
  width: 100%;
}

.c0::-webkit-input-placeholder {
  color: #cccccc;
}

.c0::-moz-placeholder {
  color: #cccccc;
}

.c0:-ms-input-placeholder {
  color: #cccccc;
}

.c0::placeholder {
  color: #cccccc;
}

.c0:focus {
  border-color: #8ce0c9;
}

.c0:-moz-read-only {
  color: #bfbfbf;
}

.c0:read-only {
  color: #bfbfbf;
}

.c1 {
  color: #737373;
  fill: currentColor;
  font-size: 0.875rem;
  height: 1rem;
  position: absolute;
  right: 0.75rem;
  top: 0.75rem;
  width: 1rem;
}

.c1 * {
  display: inline-block;
}

<div
  className=""
>
  <p>
    Text above
  </p>
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <input
      className="c0"
      maxLength={undefined}
      name="test"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      placeholder={undefined}
      readOnly={false}
      type="text"
      value=""
    />
    <div
      className="c1"
      onClick={[Function]}
      value=""
    />
  </div>
  <p>
    Text below
  </p>
</div>
`;

exports[`renders with other props 1`] = `
<div
  className="className"
>
  <label
    className="labelClassName"
    htmlFor="labelFor"
  >
    Foo
  </label>
  <p>
    Blah
  </p>
</div>
`;
